c++/mpi/window.hpp

```diff
@@ -68,83 +94,156 @@ namespace mpi {
       int target_count_ = target_count < 0 ? origin_count : target_count;
-      MPI_Put(origin_addr, origin_count, origin_datatype, target_rank, target_disp, target_count_, target_datatype, win);
-    };
+      if (has_env) {
+        MPI_Datatype origin_datatype = mpi_type<OriginType>::get();
+        MPI_Datatype target_datatype = mpi_type<TargetType>::get();
+        MPI_Put(origin_addr, origin_count, origin_datatype, target_rank, target_disp, target_count_, target_datatype, win);
+      } else {
+        if (origin_count == 0 || target_count == 0) {
+          return;
+        }
+        if ((sizeof(OriginType) * origin_count) <= (sizeof(TargetType) * target_count_)) {
+          std::memcpy(reinterpret_cast<char*>(base_local) + target_disp * sizeof(TargetType),origin_addr, sizeof(TargetType) * target_count_); // Insure correct allignment
+        } else {
+          std::abort();
+        }
+      }
+    }
 
     /// Accumulate data into target process through remote memory access.
     template <typename TargetType = BaseType, typename OriginType>
     std::enable_if_t<has_mpi_type<OriginType> && has_mpi_type<TargetType>, void>
     accumulate(OriginType const *origin_addr, int origin_count, int target_rank, MPI_Aint target_disp = 0, int target_count = -1, MPI_Op op = MPI_SUM) const noexcept {
-      MPI_Datatype origin_datatype = mpi_type<OriginType>::get();
-      MPI_Datatype target_datatype = mpi_type<TargetType>::get();
-      int target_count_ = target_count < 0 ? origin_count : target_count;
-      MPI_Accumulate(origin_addr, origin_count, origin_datatype, target_rank, target_disp, target_count_, target_datatype, op, win);
+      if (has_env) {
+        MPI_Datatype origin_datatype = mpi_type<OriginType>::get();
+        MPI_Datatype target_datatype = mpi_type<TargetType>::get();
+        int target_count_ = target_count < 0 ? origin_count : target_count;
+        MPI_Accumulate(origin_addr, origin_count, origin_datatype, target_rank, target_disp, target_count_, target_datatype, op, win);
+      } else {
+        switch (op) {
+          case MPI_SUM:
+            for (int i = 0; i < origin_count; ++i) {
+                base_local[target_disp + i] += origin_addr[i];
+            }
+            break;
+          case MPI_PROD:
+            for (int i = 0; i < origin_count; ++i) {
+               base_local[target_disp + i] *= origin_addr[i];   
+            }
+            break;
+          case MPI_MIN:
+            for (int i = 0; i < origin_count; ++i) {
+                base_local[target_disp + i] = std::min(base_local[target_disp + i], origin_addr[i]);
+            }
+            break;
+          case MPI_MAX:
+            for (int i = 0; i < origin_count; ++i) {
+                base_local[target_disp + i] = std::max(base_local[target_disp + i], origin_addr[i]);
+            }
+            break;
+          default:
+            std::cerr << "Error: Unsupported operation." << std::endl;
+            std::abort(); 
+        }
```

Comment:
For now just `std::abort`. We have to think of a usable solution. Maybe have a look at what `all_reduce` does without `has_env`.